### PR TITLE
RFC: Hacky implementation of a style guide component

### PR DESCRIPTION
**Pull request by pcardune**
In thinking about how we document styling and components to maximize the reusability of UI code, I think it's important we have an up-to-date style guide that shows off all the available components/styling and how to use them.

Here's a screenshot of what the style guide looks like:
screen shot 2016-06-13 at 3 17 19 pm

And here is a screenshot of the code you would add to insert an example into the style guide.
screen shot 2016-06-13 at 3 17 45 pm

The main benefit of this approach is that it's pretty much impossible for the style guide to go out of date because it is embedded in the code that it documents.

My main questions at the moment are:
- **Where should the style guide live?** Currently a lot of components depend on global style information, so you can't render them properly outside the context of a live page. I jammed the style guide into the admin box, but it probably shouldn't belong here.
- **Would people actually take the time to add examples for the components they create?**

What do you guys think?

### CODE REVIEW

**Comment by pcardune** [comment]
p.s. the code for this is ugly. If people like the direction, I will spend the time to make it less hacky.

**Comment by bcjordan** [comment]
I like this approach in general, especially how the examples can live alongside the component source code. In terms of placement, would prefer for the example page to live somewhere publicly accessible, though don't have a great idea for where that might be yet.

**Comment by breville** [comment]
fyi @tanyaparker .. let's chat about this more after pondering.  

We attempted a code.org/sitemap a while back to understand our content architecture, but it went out of date which this approach solves better.  

We have a number of styling issues to stay consistent on, and they range from the highly-specific (expressed in code) to the very general (philosophies to agree upon).  These include colour palette, typography, geometry (square/round/tight/loose/depth/flat/etc.), audience-variance (children/teenagers/parents/teenagers/administrators/etc), medium (web/print/video/etc.), aesthetic, and story-telling.

Regardless of where we land in these discussions, it seems very useful to express many of them in terms of example temples from which we can borrow.  Sometimes it makes sense to have a single page with examples all the individual components, and with React or partials these components can obviously be more than a single HTML element.  Sometimes it makes sense to have an entire example page, for example when the template is an entire puzzle layout, or an entire type of Pegasus page layout.

Note that we already have code.org/markdown which is the internal reference for much of our Pegasus styling.  To me, the most obvious thing we're missing on Pegasus is layout templates for complete pages that have a whole set of elements such as text & image & tables & videos all mixed together.  Currently we just have a set of paragraphs one after another, with some custom layouts for sticking pictures and videos around the place, but it would be nice to have some very lush "magazine style" templates that can be authored easily.

For this PR, if we're starting with a kind of Dashboard/React-specific version of what code.org/markdown gives us, it seems like it might be pretty useful.  Thanks for doing this.

**Comment by pcardune** [comment]
@breville I didn't know about code.org/markdown! That looks interesting.

Very nice list of styling consistency abstraction levels. Are you going to be in the office tomorrow? perhaps @tanyaparker, @breville, and I can get together to discuss?

**Comment by breville** [comment]
Yes, I'll be in.  Chat sounds perfect.

**Comment by pcardune** [comment]
Ok, moved it to a simple rails page which you can see at `/admin/styleguide`. Is this a better location or should I put it somewhere else?

Screenshot: (`lightish_teal`... really?)

screen shot 2016-06-16 at 9 57 40 am

File: dashboard/app/views/admin_styleguide/show.html.haml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#styleguide-app
+%script{src: asset_path('js/marked/marked.js')}
```
Comment (pcardune) [inline dashboard/app/views/admin_styleguide/show.html.haml:2]: I tried to include `%link{href: asset_path('css/common.css'), rel: 'stylesheet', type: 'text/css'}` in here and I got a 404, but I think it is required for some of the apps components to look right (specifically the run button).

By the way, what else should I be including in here?

**Comment by pcardune** [comment]
PTAL!

File: apps/src/templates/StyleGuide.jsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        These are the standard colors available in SCSS and javascript.
+      </p>
+      <p>
+        JavaScript usage is as follows:
+        <pre>
+          {`import colors from '@cdo/apps/colors';
+console.log("the hex value for purple is:", colors.purple);`}
+        </pre>
+      </p>
+      {Object.keys(color).map(colorKey => (
```
Comment (bcjordan) [inline apps/src/templates/StyleGuide.jsx:67]: Very cool, love the idea of this auto-updating.

File: dashboard/config/routes.rb [hunk 3]
```diff
@@ -213,6 +213,8 @@ module OPS
   post '/admin/confirm_email', to: 'admin_users#confirm_email', as: 'confirm_email'
   post '/admin/undelete_user', to: 'admin_users#undelete_user', as: 'undelete_user'
 
+  get '/admin/styleguide', :to => 'admin_styleguide#show', as: 'styleguide'
```
Comment (bcjordan) [inline dashboard/config/routes.rb:216]: Hm, good Q: re where to put it. Brainstorming a bit:
- `studio.code.org/styleguide`
- `studio.code.org/dev/styleguide`
- `studio.code.org/test/styleguide`
- `studio.code.org/about/styleguide`

**Comment by bcjordan** [comment]
Just realized we will also be able to put this under an eyes screenshot so we would see if any low-level UI components change.

**Comment by breville** [comment]
This all seems like goodness to me.  It makes sense to start trawling for other standalone components that we have, including dialogs (bonus points if you can have them render in a page without actually popping up over it), headers & body text, the various course_blocks, and anything else that is already componentized.

**Comment by pcardune** [comment]
@breville so is that an "LGTM" from you and I should go ahead and merge?

**Comment by breville** [comment]
Yes.  I don't know React much so can't comment much on the code itself, but the outcome looks good, and doesn't seem like it will interfere with anything else on the site.

**Comment by pcardune** [comment]
Closing this to merge into staging-next isntead

**Pull request closed without merging**